### ISSUE TITLE
Inference speed decreasing with batch size (PyTorch)

### ISSUE THREAD

**Issue by tshindmarsh** [DeepLabCut/DeepLabCut#2845]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Bug description

Hi!

First of all, thanks for all your continued work on developing and improving DeepLabCut! Very excited to be testing out 3.0!

I have unfortunately run into a slight issue running the latest version on my cluster. Training proceeds just fine (took about 12hrs for 300 epochs), but inference has been painfully slow despite the GPU being engaged. My videos are quite large at 2048x2048, and I am trying to track 4 small individual animals that each occupy about 200px. For a 30 minute video, inference is taking upwards of 6 hours with batch size 32. Curiously, however, the inference speed improves about 4x by reducing the batch size to 1. Any batch size above 1 progressively decreases inference speed. The tracking itself looks good, though!

I am running with a 48GB NVIDIA L40S as my GPU and 16 CPU cores, 128GB RAM. nvidia-smi clearly shows the GPU being engaged, with memory usage pinned around 21893/46068MiB. However, when batch size goes above 1, the volatile GPU utilization begins to oscillate between 0% and 80-90% every few seconds. I suspect this may perhaps be the source of my issue, but I really have no idea how to best address the issue. Do you have any suggestions on how to best proceed? 

### Operating System

Linux (cluster)

### DeepLabCut version

dlc version DLC 3.0.0rc6

### DeepLabCut mode

multi animal

### Device type

48GB NVIDIA L40S

### Steps To Reproduce

DLC and PyTorch both installed in a conda environment using the guide here: https://deeplabcut.github.io/DeepLabCut/docs/pytorch/user_guide.html

torch.cuda.is_available() returns TRUE and torch reports the correct device

Python code: 

```python
import deeplabcut
import os
import torch

confnameC='XXX/config.yaml'

videoFolder = 'XXX/testVids'
videoNames = [os.path.join(videoFolder, f) for f in os.listdir(videoFolder)
              if f.endswith('.mp4') and not f.endswith('_labeled.mp4')]

torch_kwargs = {
    "device": "cuda:0" 
}

deeplabcut.analyze_videos(
    confnameC,
    videoNames,
    videotype='mp4',
    shuffle=1,
    trainingsetindex=0,
    save_as_csv=True,
    batchsize=16,
    auto_track=True,
    **torch_kwargs,
)
deeplabcut.filterpredictions(confnameC, videoNames, videotype='.mp4')
deeplabcut.create_labeled_video(
    confnameC,
    videoNames,
    videotype='.mp4',
    filtered=True,
    draw_skeleton = True,
    keypoints_only=False,
    color_by='individual',
)
```

### Relevant log output

```shell
Analyzing videos with XXX/train/snapshot-best-300.pt
Starting to analyze XXX/testVids/Untitled.mp4
Video metadata: 
  Overall # of frames:    89999
  Duration of video [s]:  1800
  fps:                    50.00
  resolution:             w=2048, h=2048

Running pose prediction with batch size 16
```

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by n-poulsen** [comment]
Hi @tshindmarsh! Thanks for reporting this.

Increasing the batch side might not always lead to a speed-up during video inference, as other operations can become the bottleneck (rather than the model's forward pass itself). So it's possible that when running on such large videos, reading the frames and "stacking" them to create a batch is becoming more expensive than the amount of time gained running the forward pass of the model.

I'll try to look into this and see if I can reproduce this issue, and whether loading frames from a video can be sped up.

**Comment by maximpavliv** [comment]
Hello @tshindmarsh ,

Sorry for the very late answer.

What is your project configuration (can you share the content of your _config.yaml_)?
Also what model architecture are you using? Can you share your _pytorch_config.yaml_ as well?

**Comment by tshindmarsh** [comment]
Hi @maximpavliv, 

I am running using a top-down approach, with a fasterrcnn_mobilenet_v3_large_fpn followed by a ResNet50. This works quite well because my animals are very small compared to the frame size. 

I did also try this on a single-animal project with just a PyTorch ResNet, and had similar issues with inference time increasing dramatically with any batch size larger than 1. It seems like frames get processed in bursts based on the GPU behavior, with the GPU utilization rising and then stalling before starting over. The duration of the stall increases with larger batch sizes. The problem is ameliorated if I downsample the videos by 50% to 1024x1024!

**Project Config**
```
# Project definitions (do not edit)
Task: MultiCricketTracking
scorer: THS
date: Oct4
multianimalproject: true
identity: false


# Project path (change when moving around)
project_path: XX/MultiCricketTrackingSherlock-THS-2024-10-04_runSherlock


# Default DeepLabCut engine to use for shuffle creation (either pytorch or tensorflow)
engine: pytorch


# Annotation data set configuration (and individual video cropping parameters)
video_sets:
  XX.mp4:
    crop: 0, 2048, 0, 2048
  XX.mp4:
    crop: 0, 2048, 0, 2048
  XX.mp4:
    crop: 0, 2048, 0, 2048
  XX.mp4:
    crop: 0, 1824, 0, 1824
  XX.mp4:
    crop: 0, 2048, 0, 2048
  XX.mp4:
    crop: 0, 1824, 0, 1824
  XX.mp4:
    crop: 0, 1824, 0, 1824
  XX.mp4:
    crop: 0, 1824, 0, 1824
  XX.mp4:
    crop: 0, 1824, 0, 1824
    
individuals:
- individual1
- individual2
- individual3
- individual4
uniquebodyparts: []
multianimalbodyparts:
- head
- abdomen
- genitalia
bodyparts: MULTI!


# Fraction of video to start/stop when extracting frames for labeling/refinement
start: 0
stop: 1
numframes2pick: 30

# Plotting configuration
skeleton:
- - head
  - abdomen
- - abdomen
  - genitalia
skeleton_color: white
pcutoff: 0.2
dotsize: 3
alphavalue: 0.7
colormap: brg


# Training,Evaluation and Analysis configuration
TrainingFraction:
- 0.95
iteration: 0
default_net_type: dlcrnet_ms5
default_augmenter: multi-animal-imgaug
default_track_method: ellipse
snapshotindex: -1
detector_snapshotindex: -1
batch_size: 16

# Cropping Parameters (for analysis and outlier frame detection)
cropping: false
#if cropping is true for analysis, then set the values here:
x1: 0
x2: 640
y1: 277
y2: 624


# Refinement configuration (parameters from annotation dataset configuration also relevant in this stage)
corner2move2:
- 50
- 50
move2corner: true

# Conversion tables to fine-tune SuperAnimal weights
SuperAnimalConversionTables:
detector_batch_size: 1
```

**Pytorch Config**
```
data:
  colormode: RGB
  inference:
    normalize_images: true
    top_down_crop:
      width: 150
      height: 150
  train:
    affine:
      p: 0.5
      rotation: 30
      scaling:
      - 0.8
      - 1.2
      translation: 0
    collate:
    covering: false
    gaussian_noise: 12.75
    hist_eq: false
    motion_blur: false
    normalize_images: true
    top_down_crop:
      width: 150
      height: 150
detector:
  data:
    colormode: RGB
    inference:
      normalize_images: true
    train:
      affine:
        p: 0.5
        rotation: 30
        scaling:
        - 0.8
        - 1.2
        translation: 40
      collate:
        type: ResizeFromDataSizeCollate
        min_scale: 0.4
        max_scale: 1.2
        min_short_side: 128
        max_short_side: 1152
        multiple_of: 32
        to_square: false
      hflip: true
      normalize_images: true
  device: auto
  model:
    type: FasterRCNN
    freeze_bn_stats: true
    freeze_bn_weights: false
    variant: fasterrcnn_mobilenet_v3_large_fpn
  runner:
    type: DetectorTrainingRunner
    key_metric: test.mAP@50:95
    key_metric_asc: true
    eval_interval: 10
    optimizer:
      type: AdamW
      params:
        lr: 0.0001
    scheduler:
      type: LRListScheduler
      params:
        milestones:
        - 160
        lr_list:
        - - 1e-05
    snapshots:
      max_snapshots: 5
      save_epochs: 20
      save_optimizer_state: false
  train_settings:
    batch_size: 16
    dataloader_workers: 1
    dataloader_pin_memory: true
    display_iters: 500
    epochs: 200
device: auto
metadata:
  project_path: 
    /oak/stanford/groups/lluo/Tom/DeepLabCut/DLC3/Models/MultiCricketTrackingSherlock-THS-2024-10-04
  pose_config_path: 
    /oak/stanford/groups/lluo/Tom/DeepLabCut/DLC3/Models/MultiCricketTrackingSherlock-THS-2024-10-04/dlc-models-pytorch/iteration-0/MultiCricketTrackingOct4-trainset95shuffle1/train/pytorch_config.yaml
  bodyparts:
  - head
  - abdomen
  - genitalia
  unique_bodyparts: []
  individuals:
  - individual1
  - individual2
  - individual3
  - individual4
  with_identity: false
method: td
model:
  backbone:
    type: ResNet
    model_name: resnet50_gn
    output_stride: 16
    freeze_bn_stats: false
    freeze_bn_weights: false
  backbone_output_channels: 2048
  heads:
    bodypart:
      type: HeatmapHead
      weight_init: normal
      predictor:
        type: HeatmapPredictor
        apply_sigmoid: false
        clip_scores: true
        location_refinement: true
        locref_std: 7.2801
      target_generator:
        type: HeatmapGaussianGenerator
        num_heatmaps: 3
        pos_dist_thresh: 17
        heatmap_mode: KEYPOINT
        gradient_masking: true
        background_weight: 0.0
        generate_locref: true
        locref_std: 7.2801
      criterion:
        heatmap:
          type: WeightedMSECriterion
          weight: 1.0
        locref:
          type: WeightedHuberCriterion
          weight: 0.05
      heatmap_config:
        channels:
        - 2048
        kernel_size: []
        strides: []
        final_conv:
          out_channels: 3
          kernel_size: 1
      locref_config:
        channels:
        - 2048
        kernel_size: []
        strides: []
        final_conv:
          out_channels: 6
          kernel_size: 1
net_type: resnet_50
runner:
  type: PoseTrainingRunner
  gpus:
  key_metric: test.mAP
  key_metric_asc: true
  eval_interval: 10
  optimizer:
    type: AdamW
    params:
      lr: 0.0001
  scheduler:
    type: LRListScheduler
    params:
      lr_list:
      - - 1e-05
      - - 1e-06
      milestones:
      - 160
      - 190
  snapshots:
    max_snapshots: 5
    save_epochs: 25
    save_optimizer_state: false
train_settings:
  batch_size: 16
  dataloader_workers: 1
  dataloader_pin_memory: true
  display_iters: 500
  epochs: 200
  seed: 42
```

**Comment by MMathisLab** [comment]
@tshindmarsh can you see if rc10 (the latest release) helps with this issue?

**Comment by maximpavliv** [comment]
Hi @tshindmarsh !

Thanks so much for raising this issue, and even more for the very detailed report you shared — it was incredibly helpful 🙏. Sorry it has taken a while to get back to you with a proper fix!

We’ve recently identified the root cause and merged a solution (PR https://github.com/DeepLabCut/DeepLabCut/pull/3099). The problem was in how batching was implemented: when building each batch, frames were being concatenated in a way that caused repeated **Tensor copies**. This meant that:

- Larger batch sizes actually led to more _copying work_ rather than efficiency gains — so inference speed went down with batch size.

- The GPU utilization oscillations you noticed (0% → 80–90%) came from the fact that most time was being spent in CPU-side data preparation, not actual GPU inference.

This is particularly noticeable with high-resolution videos like yours (2048×2048), where copying becomes very expensive.

The fix is included in our new release **DLC 3.0rc12**, where batching is now handled without repeated copying. This should resolve the slowdown you observed.

Thanks again for your patience and for sharing your experience — community feedback like this really helps us improve DeepLabCut. We’re also working on further optimizations to inference speed, so you can expect more improvements in upcoming releases.

Since the specific issue of inference speed decreasing with batch size should now be solved, I’ll go ahead and close this issue. Please feel free to re-open if you still notice unexpected behavior after updating!

💜 The DeepLabCut Team

**Comment by HagarYulzari30** [comment]
Hi, sorry to jump into the thread :)

First, I want to say that I’m very happy with the model’s detection accuracy, and I really appreciate the continuous improvements to the program.

I’m posting because I’m currently using DLC 3.0rc13 and still experiencing behavior very similar to what was described above.

My videos are also quite large, approximately 2 hours long at 1040×1440 resolution and 100 FPS. 
I’m running inference on an NVIDIA RTX 4090 using a ResNet50 model with compile, autocast, and multithreading enabled. I tried to increase the queue_length to 128 and 256 which helped a bit but not significantly.

With batch_size=1, I reach around 65–68 FPS. However, when increasing the batch size, I start seeing the same GPU and FPS fluctuations described in this thread and across batch sizes 4–64, the average throughput remains around 45 FPS.

Since this issue was mentioned as fixed in rc12 (which I also tested and observed similar behavior with), I’d really appreciate any insight into whether I might be missing something on my side or if there are additional settings worth investigating.

I process a very large dataset (about 30 videos per day), so inference speed is a critical factor for me.

I’m attaching my config.yaml and pytorch_config.yaml files below.

Thanks so much!


[config.yaml](https://github.com/user-attachments/files/27513995/config.yaml)
[pytorch_config.yaml](https://github.com/user-attachments/files/27513996/pytorch_config.yaml)

**Comment by C-Achard** [comment]
Hello @HagarYulzari30,

Thanks for sharing and posting in the relevant thread; if I'm not mistaken we have very fresh speed benchmarking results which imply that the speed gains from batch size sharply decrease as resolution increases, which given your resolution is most likely what is happening.
For low resolutions increasing batch size multiplies the speed by a noticeable factor, but this gain becomes less and less noticeable as resolution increases, and may even decrease performance at some point.

<img width="938" height="406" alt="Image" src="https://github.com/user-attachments/assets/674aa0b0-915d-4824-a7a4-f11392f634a6" />


We will be looking into it, alongside other quality control efforts for the 3.0 release, but I cannot give an exact timeline for this particular problem for now.
One advice I can give while we investigate: could you simply use lower resolution data for the model, see whether performance remains acceptable, and convert keypoints coordinates back to the high-res data ?
As long as the down/upsampling factor is known this should be relatively easy, though downsampling will take time of its own.

**Comment by HagarYulzari30** [comment]
Hey @C-Achard ,
Thank you so much for the prompt response, and sorry for my delayed reply! 
This was very useful information, and it also explains why my previous attempts to crop the video led to almost no change in speed (I cropped it by half, but the resolution still stayed above 512x512 pixels).

I tried downscaling the video so the horizontal size fits 500 px, and it actually boosted the FPS to well above 400! Nonetheless, it also reduced the model confidence, and the mean likelihood dropped below 0.6 for most body parts.

But the good news for me is that my main area of interest, which mainly includes the rats’ torso movements during the task, is limited to roughly a 500×500 px region. So instead of doing such a massive downscale that increases the noise in the trajectories, I can just use the cropping option in DLC. This keeps the model performance much higher while still bringing the inference speed up to around 250 FPS.

So thanks again for the help, I think I found a good solution for the meantime and I’ll be looking forward to the new DLC updates!

**Comment by C-Achard** [comment]
@deruyter92 Closing again for now, if there is more to add here feel free to re-open
